client/concordclient/include/client/concordclient/trc_queue.hpp

```diff
@@ -0,0 +1,112 @@
+// Concord
+//
+// Copyright (c) 2020-2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <condition_variable>
+#include <boost/lockfree/spsc_queue.hpp>
+#include <list>
+#include <memory>
+#include <thread>
+#include <mutex>
+#include <exception>
+#include <chrono>
+#include <variant>
+
+#include "assertUtils.hpp"
+#include "event_update.hpp"
+
+namespace concord::client::concordclient {
+
+typedef std::variant<Update, EventGroup> EventVariant;
+
+class TrcQueue {
```

Comment:
Can we implement this as :
template \<typename T\>
class TrcQueue : public IQueue\<T\> 

So TrcQueue will be another implementation of IQueue

Reply:
@arc-vmware if `TrcQueue` derives from `IQueue`, there will be a type mismatch b/w the `push` method of TrcQueue and IQueue, as the latter requires `std::unique_ptr<T>` as an argument, whereas with spsc queue we can only use raw pointers. 
This is because spsc_queue does not use move semantics (methods to push onto the queue take parameters as const reference and construct a copy in the queue's internal storage), so i can not push a unique_ptr<T> onto the TrcQueue, i.e., why I used a separate implementation.

Reply:
void push(EventVariant* update) {
    queue_data_.push(std::move(update));
    condition_.notify_one();
  }

Here we can see that we use move semantics. I think it will be better if we refactor and use smart ptrs instead of raw ptr. Also, we can pass the ptr inside a unique_ptr as below:
auto update = std::unique_ptr<EventVariant>{raw_ptr_update};
remote_queue->push(std::move(update));
Internally the queue will use the unique_ptr. I think enforcing the queue to use a smart ptr (unique_ptr in this case) is a better design. It will be good if we give it a thought and try it. If it doesn't work we can do things as it is.

Reply:
I agree, but boost::lockfree::spsc_queue doesn't allow it. Please see https://www.boost.org/doc/libs/1_59_0/doc/html/boost/lockfree/spsc_queue.html, this page mentions boost::lockfree::spsc_queue of type T has the following requirements -
- T must have a default constructor
- T must be copyable
A unique_ptr type is not copyable.